### ISSUE TITLE
[ssh] SSH support in UCloud

### ISSUE THREAD

**Issue by DanThrane** [SDU-eScience/UCloud#3647]
__Goal 1:__ To enable access to workloads via publicly accessible SSH. This will allow users to use a vast amount of
tools which require SSH. Benefits include: file transfer (SCP, SSHFS, rsync), access to workloads via a personal
terminal and port-forwarding to arbitrary ports.

__Goal 2:__ Make public SSH keys a first class citizen in UCloud. This is useful for a variety of different workloads
in UCloud. Examples include: virtual machine workloads, traditional HPC systems which provider SSH access and access
to supported container workloads.

## Summary

1. Allow end-users to upload public SSH keys into UCloud
2. Allow applications to declare that they are SSH enabled (this will be opt-in)
3. For every SSH enabled job, allocate a port on `ssh.cloud.sdu.dk` for the duration of the job
4. The allocated port is forwarded by the gateway to a matching internal IP address
5. The internal IP address will forward traffic into the user job (where an SSH server must be running)

### Implement public keys as a resource in UCloud

Expand the "Resources" side-panel with a new tab for SSH keys. Here the user should be able to upload SSH keys. The
backend should perform basic validation and reject anything which looks like a private key.

UCloud/Core should store these keys in its own datastore. These keys should automatically be made available to any
provider when they run an SSH enabled workload.

#### Proposed endpoints

```kotlin
// NOTE(Dan): The API below mirrors the conventions used by the normal resource system of UCloud without actually
// being part of the resource system.

data class SSHKey(
    @UCloudApiDoc("An opaque and unique identifier representing this SSH key")
    val id: String,
    @UCloudApiDoc("The UCloud username of the user who owns this key")
    val owner: String,
    @UCloudApiDoc("Timestamp for when this key was created in UCloud")
    val createdAt: Long,
    val specification: Spec
) {
    data class Spec(
        @UCloudApiDoc("Single line human-readable description, not unique")
        val title: String,
        @UCloudApiDoc("The public part of an SSH key")
        val key: String,
    )
}

// End-user API
// ===================================================================================================================
// Jobs.uploadSshKey: RpcCall<BulkRequest<SSHKey.Spec>, BulkResponse<FindByStringId>>
//   Uploads an SSH key to UCloud/Core the response contains the generated IDs for the public keys.
// 
// Jobs.retrieveSshKey: RpcCall<FindByStringId, SSHKey>
//   Retrieves a concrete SSH key by an ID. Only the end-user who owns the key is allowed to read it back.
// 
// Jobs.browseSshKeys: RpcCall<PaginationRequestV2, PageV2<SSHKey>>
//   Paginated API for retrieving SSH keys
// 
// Jobs.deleteSshKeys: RpcCall<BulkRequest<FindByStringId>, Unit>
//   Deletes one or more SSH keys by their unique identifier. Only the user who owns the key is able to delete it.

// Provider API
// ===================================================================================================================
// JobsControl.browseSshKeys: RpcCall<FindByStringId + PaginationRequestV2, PageV2<SSHKey>>
//   Paginated API which allows a provider to find all relevant SSH keys for a Job identified by its ID.
// 
//   TODO: Which keys are relevant for a job? Most likely this should return the keys of every user who has "EDIT"
//   rights for the job (i.e. every user capable of cancelling the job).
//
//   NOTE: It is very important that pagination is stable for the provider API (as opposed to the user API, where it is
//   just desireable). This API should likely use an monotonically increasing identifier for sorting and use this key
//   to determine the `next` token.
```

### Allow applications to declare that they are SSH enabled

SSH enabled applications are applications which automatically start an SSH server. Supported providers must be capable
of configuring the SSH server with the keys retrieved from the provider API (see previous section).

This will expand application definitions with the following section:

```yaml
ssh:
    enabled: true # default value is false
```

This section might be expanded later. In that case, the default behavior will not change. SSH enabled applications
should still be accessible without having any SSH keys uploaded to UCloud. In that case, users can only access the
application through other means (e.g. "Open terminal" or "Open interface").

#### Container applications

The following behavior must be implemented for container applications which needs to be SSH enabled. Container
applications are _any application currently in production, which is not a virtual machine_.

1. The user must be named `ucloud` with UID 11042
2. The user's home directory must be configured as `/home/ucloud`
3. The application must automatically start an SSH server
  - This should happen even if the user has no SSH keys uploaded
4. The SSH server must run on port 22 and listen on all interfaces
5. The SSH server must have a reasonably secure configuration (Martin should probably lead this)
  - Password logins should not be allowed

The provider will mount relevant keys in `/home/ucloud/.ssh/authorized_keys` shortly after an SSH enabled application 
starts. The provider will not change these keys after application startup. The only way to add/remove keys is to
restart the application. This behavior depends on the compute backend, but it will be the expected behavior for
container backends.

### Implement support for SSH enabled applications in UCloud compute plugin

This task will allow the UCloud compute plugin (previously known as `app-kubernetes`) to support SSH enabled
applications. This implementation requires support from the infrastructure and is based largely on the work we have
already done for public IPs.

The plugin automatically assigns an available port to SSH enabled applications. The plugin may choose to only do this
if the end-user has uploaded any relevant keys to UCloud/Core. Public keys are installed as described in the previous
section. Once a port has been allocated, it will send a message back to the end-user with instructions on how to 
connect to the job. 

We may create a specialized UI for displaying this information. For example, we might add a new card on the "Runs"
page which looks roughly like this (In this case `2032` is the allocated port number):

```
+---------------------------------------------------------+
| SSH access                                      (Help)  |
| ------------------------------------------------------- |
| This application supports access via SSH. You may use   |
| the following command to access it:                     |
|                                                         |
| ssh ucloud@ssh.cloud.sdu.dk -p 2032             (Copy)  |
|                                                         |
+---------------------------------------------------------+
```

Behind the scenes, this solution is identical to the public IP solution, but using a slightly different mechanism for
mapping. In the public IP solution, the plugin performs a mapping from a public IP address to an internal IP address.
This internal IP address is mounted on the host (and inside of the container). Traffic is forwarded from the gateway
to the appropriate internal IP address. For SSH access, we will be mapping an allocated port to an internal IP
address. The gateway will forward traffic addressed to `(ssh.cloud.sdu.dk, $PORT)` to a matching 
`(mapToPortIp($PORT), 22)`. The `mapToPortIp` procedure will use a process similar to:

```c
#include <stdio.h>

int main(int argc, char *argv[])
{
    for(int p = 2032; p < 4064; p++)
    {
        int n = 1+p/254;
        int m = 1+p%254;
        printf("%d.%d\n", n, m);
    }
    return 0;
}
```

If the internal subnet allocated for SSH ports is `10.42.n.m` then UCloud would mount `10.42.9.1` for the
job described above.

**Comment by Tobias-V** [comment]
> Allow applications to declare that they are SSH enabled (this will be opt-in)

> - 
> -     The application must automatically start an SSH server
> - 
> -     This should happen even if the user has no SSH keys uploaded
> - 
> -     The SSH server must run on port 22 and listen on all interfaces

Would it be possible to implement this in a way where this can be toggled when submitting the job?
Perhaps by feeding an option to the start up script of the application that installs and configures the server.

I'm concerned that the would be reluctant to SSH enable apps if it's going to impact users of the app that don't need the feature.

**Comment by DanThrane** [comment]
> Would it be possible to implement this in a way where this can be toggled when submitting the job? Perhaps by feeding an option to the start up script of the application that installs and configures the server.

Yes, I'll add an option to disable SSH for applications which are enabled.

**Comment by Tobias-V** [comment]
Wouldn't it have to be the other way around to avoid adding unnecessary bloat to the images?

**Comment by DanThrane** [comment]
The application YAML already need expliclty enable it. This section

```
ssh:
    enabled: true # default value is false
```

goes into the application YAML. If it is not present (or if `enabled: false`) then the application won't be expected to implement an SSH server.
